README.md
# rest-collection-stream

A simple HTTP client for streaming JSON content from REST lists and collections. Includes extensive pagination support for continuous streaming across multiple HTTP requests.

Use this module to pull an entire collection of JSON documents from your favorite API into a Node.js stream! It will parse the response as JSON, extract content as objects, push it down the stream, and continue doing so as long as there is more data.

```js
var restCollection = require('rest-collection-stream');
restCollection('https://somecloudapp.io/api/v1/resource').pipe(/*somewhere*/);
```

## Installation

```bash
npm install rest-collection-stream
```

## Documentation

The module consists of a function returning a stream.

It's that simple.

Pulling data from an API is much more fun with streaming! Especially when paging is handled automatically...

```js
var restCollection = require('rest-collection-stream');
var fs = require('fs');
var es = require('event-stream');

// export all customers into a backup file
restCollection('https://somecloudapp.io/api/v1/customers')
  .pipe(es.map(/* my mapping function */))
  .pipe(es.stringify())
  .pipe(fs.createWriteStream(backupFile));
```

More precisely, the function has the same signature as [request](https://github.com/mikeal/request), accepting the same [parameters and options](https://github.com/mikeal/request/blob/master/README.md#requestoptions-callback), but returning a stream instead of invoking a callback.

```js
var stream = restCollection(url, options);
```

In addition to request options, you may also set:
- `data` - a function for extracting content from the response
- `next` - a function for handling pagination

The following options are also enforced:
- `json` - always **true**, only JSON APIs are supported at the moment and the response body is always parsed as such
- `headers` - HTTP compression is enabled by default via `Accept-Encoding` - a nicety if you are pulling thousands of documents from the API servers

### Streaming objects from the response

By default, if the response body consists of an array, all elements are emitted down the stream.

If the response is an envelope object, commonly used properties will be inspected for an array of data. Those are (in order): `data`, `results`, `items`, `value` and `objects`.

#### Custom parsing

If the elements you want to stream are not found in those common properties, you can supply your own function to extract the elements. Your function is expected to return an array of items to emit down the stream.

To use a custom function for parsing the body, specify it as `data` in the options.

```js
restCollection(url, {
  data: function (res, body) {
    return body.items;
  }
}).pipe(...);
```

#### Custom streaming

You can also gain full control over what gets emitted by not returning values from your `data` function. Use `this.emit('data', item)` to send whatever you want, including results from an asynchronous function.

```js
restCollection(url, {
  data: function (res, body) {
    this.emit('data', { something: 'else' });
  }
}).pipe(...);
```

### Pagination

Another request for the next page of data is automatically sent if there is more data available. The module implements common pagination support from popular APIs, but you can also supply your own.

#### Built-in pagination support

- Standard HTTP header `Link: <URL>; rel='next'`
- Common envelope properties
    + Facebook: `paging.next`
    + Google APIs: `nextLink`
    + OData 2.0 and 3.0: `__next`
    + OData 4.0: `@odata.nextLink`
    + AWS: `NextMarker`
- A self-incrementing `page` query string parameter

Should I support more APIs? Pull requests are welcome.

#### Custom paging

If your API works differently, you can provide you own paging function. The function can return the URL of the next page or query string parameters to apply on the next request. To stop pagination, return something falsy.

To use a custom function for pagination, specify it as `next` in the options.

```js
restCollection(url, {
  next: function (res, body) {
    return urlOfNextPage;
  }
}).pipe(...);
```

To specify the URL of the next page to request, return it as a string. The URL is automatically resolved using [url.resolve](http://nodejs.org/api/url.html#url_url_resolve_from_to).

To specify query string parameters instead, return an object. The key/value pairs will be applied on the next request. The current query string values can be accessed in `this.req.qs`.

```js
restCollection(url, {
  next: function (res, body) {
    // increment the page manually (we should also have a stop condition...)
    return { page: this.req.qs.page + 1 };
  }
}).pipe(...);
```

To use anything else, such as request headers, you may also access the request parameters using `this.req`. Any modification will be used onward for the next request. Make sure to return the URL, otherwise paging will stop.

```js
restCollection(url, {
  next: function (res, body) {
    // return the same URL - paging is controlled using headers instead
    this.req.headers['X-Page-Index'] = nextPage;
    return this.req.uri.href;
  }
}).pipe(...);
```

## Dependencies

+ [request](https://github.com/mikeal/request)
+ [event-stream](https://github.com/dominictarr/event-stream)

## License

The MIT License (MIT)

Copyright (c) 2014, Nicolas Mercier

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


HISTORY.md
## 0.0.1 - May 1st, 2014
Initial release


test/test-rest-client.js
require('should');
var restCollection = require('../');
var es = require('event-stream');

describe('restCollection API', function () {
  it('should return a stream', function () {
    restCollection.request = function (options, callback) {
      callback(null, {}, []);
    };
    var stream = restCollection('http://host/someapi');
    stream.should.have.properties('pipe', 'pause', 'resume');
    stream.pipe.should.be.a.Function;
    stream.pause.should.be.a.Function;
    stream.resume.should.be.a.Function;
  });
  it('should apply options', function () {
    restCollection.request = function (options, callback) {
      options.uri.href.should.eql('http://host/someapi');
      options.qs.should.eql({ $top: 100 });
      options.headers.should.containEql({
        'Accept': 'application/json',
        'Accept-Encoding': 'gzip, deflate',
        'X-Header': 'value'
      });
      options.should.containEql({ someOption: true });
      options.json.should.be.true;
      options.data.should.be.a.Function;
      options.next.should.be.a.Function;
    };
    restCollection('http://host/someapi', {
      qs: { $top: 100 },
      headers: { 'X-Header': 'value' },
      someOption: true
    });
  });
  it('should support stream piping', function () {
    restCollection.request = function (options, callback) {
      callback(null, {}, [1, 2, 3, 4, 5]);
    };
    restCollection('http://host/someapi')
      .pipe(es.join(','))
      .pipe(es.wait(function (err, text) {
        text.should.be.exactly('1,2,3,4,5');
      }));
  });
  it('should parse data when response body has an envelope', function (done) {
    restCollection.request = function (options, callback) {
      callback(null, {}, {
        meta: { important: true },
        data: [1, 2, 3, 4, 5]
      });
    };
    restCollection('http://host/someapi')
      .pipe(es.join(','))
      .pipe(es.wait(function (err, text) {
        text.should.be.exactly('1,2,3,4,5');
      }))
      .on('end', done);
  });
  it('should parse data with a custom function', function (done) {
    restCollection.request = function (options, callback) {
      callback(null, {}, {
        very: {
          unusual: [
            { data: [1, 2, 3, 4, 5] }
          ]
        }
      });
    };
    function unusualParser(res, body) {
      return body.very.unusual[0].data;
    }
    restCollection('http://host/someapi', { data: unusualParser })
      .pipe(es.join(','))
      .pipe(es.wait(function (err, text) {
        text.should.be.exactly('1,2,3,4,5');
      }))
      .on('end', done);
  });
  it('should paginate with the "Link" header', function (done) {
    restCollection.request = function (options, callback) {
      if (options.uri.path === '/someapi/page1') {
        callback(null, {
          headers: { 'Link': '<http://host/someapi/page2>; rel=\'next\'' }
        }, {
          data: [1, 2, 3]
        });
      } else if (options.uri.path === '/someapi/page2') {
        callback(null, {
          headers: { 'Link': '<http://host/someapi/page3>; rel=\'next\'' }
        }, {
          data: [4, 5, 6],
        });
      } else {
        callback(null, {
          headers: {}
        }, {
          data: [7, 8]
        });
      }
    };
    restCollection('http://host/someapi/page1')
      .pipe(es.join(','))
      .pipe(es.wait(function (err, text) {
        text.should.be.exactly('1,2,3,4,5,6,7,8');
      }))
      .on('end', done);
  });
  it('should paginate with common body fields', function (done) {
    restCollection.request = function (options, callback) {
      if (options.uri.path === '/someapi/page1') {
        callback(null, {}, {
          data: [1, 2, 3],
          paging: { next: '/someapi/page2' }
        });
      } else if (options.uri.path === '/someapi/page2') {
        callback(null, {}, {
          data: [4, 5, 6],
          nextLink: 'http://host/someapi/page3'
        });
      } else {
        callback(null, {}, { data: [7, 8] });
      }
    };
    restCollection('http://host/someapi/page1')
      .pipe(es.join(','))
      .pipe(es.wait(function (err, text) {
        text.should.be.exactly('1,2,3,4,5,6,7,8');
      }))
      .on('end', done);
  });
  it('should paginate with the query string', function (done) {
    var data = [1, 2, 3, 4, 5, 6, 7, 8];
    restCollection.request = function (options, callback) {
      var index = (+options.qs.page - 1) * 3;
      callback(null, {}, data.slice(index, index + 3));
    };
    restCollection('http://host/someapi', { qs: { page: 1 }})
      .pipe(es.join(','))
      .pipe(es.wait(function (err, text) {
        text.should.be.exactly('1,2,3,4,5,6,7,8');
      }))
      .on('end', done);
  });
  it('should paginate with a custom function', function (done) {
    var data = [1, 2, 3, 4, 5, 6, 7, 8, 9];
    restCollection.request = function (options, callback) {
      var index = (+options.qs.$page_index || 0) * 3;
      callback(null, {
        headers: {
          'X-Page-Index': options.qs.$page_index || 0,
          'X-Page-Count': 3
        }
      }, data.slice(index, index + 3));
    };
    function customPager(res, body) {
      var index = +res.headers['X-Page-Index'];
      var count = +res.headers['X-Page-Count'];
      if (index < count - 1) {
        // set query string parameter computed from response headers
        return { $page_index: index + 1 };
      }
      return null;
    }
    restCollection('http://host/someapi', { next: customPager })
      .pipe(es.join(','))
      .pipe(es.wait(function (err, text) {
        text.should.be.exactly('1,2,3,4,5,6,7,8,9');
      }))
      .on('end', done);
  });
});



index.js
var request = require('request');
var zlib = require('zlib');
var parseUrl = require('url').parse;
var resolveUrl = require('url').resolve;
var es = require('event-stream');

var isUrl = RegExp.prototype.test.bind(/^http(s)?:/);
var parseLink = RegExp.prototype.exec.bind(/<(.+?)>;\s*rel=['"]?next['"]?/);

function isObject(obj) {
  return obj === Object(obj);
}

var defaults = {
  data: function (res, body) {
    // body is an array, return as-is
    if (Array.isArray(body)) return body;
    // body is an envelope, look for commonly used properties
    var keys = ['data', 'results', 'items', 'value', 'objects'];
    for (var i = 0, len = keys.length; i < len; i++) {
      if (!(keys[i] in body)) continue;
      var value = body[keys[i]];
      if (Array.isArray(value)) return value;
    }
    // unable to detect collection items, need to provide your own function
    return null;
  },
  next: function (res, body) {
    // inspect the "Link" header
    if (res.headers) {
      var match = parseLink(res.headers.Link || res.headers.link);
      if (match) return match[1];
    }
    // look for commonly used properties
    if (body.paging && body.paging.next) return body.paging.next;
    if (body.meta && body.meta.next) return body.meta.next;
    //   Google APIs
    if (typeof body.nextLink === 'string') return body.nextLink;
    //   OData 2.0 and 3.0
    if (typeof body.__next === 'string') return body.__next;
    //   OData 4.0
    if (body['@odata.nextLink']) return body['@odata.nextLink'];
    //   AWS
    if (typeof body.NextMarker === 'string') {
      return { Marker: body.NextMarker };
    }
    // inspect the query string
    if (typeof this.req.qs.page === 'number') {
      // increment page if current page has data
      var data = this.req.data(res, body);
      if (data && data.length) {
        return { page: this.req.qs.page + 1 };
      }
    }
    // need to provide your own function (should we look for more?)
    return null;
  }
};

var restCollection = function (url, o) {

  if (isObject(url)) {
    o = url;
    url = null;
  }

  var options = {
    uri: url ? parseUrl(url) : null,
    qs: {},
    headers: {},
    json: true,
    data: defaults.data,
    next: defaults.next
  };

  o && Object.keys(o).forEach(function (key) {
    options[key] = o[key];
  });

  // setup headers - only JSON is supported
  options.headers = options.headers || {};
  options.headers.Accept = 'application/json';
  options.headers['Accept-Encoding'] = 'gzip, deflate';

  return es.readable(function (count, callback) {
    // nowhere to go next, end the stream
    if (!options.uri) return this.emit('end');

    var req = restCollection.request || requestWithEncoding;
    var stream = this;
    options.pageIndex = count;
    stream.req = options;

    req(options, function (err, res, body) {
      if (err) return stream.emit('error', err);
      stream.emit('page', options, res, body);

      var data = options.data.call(stream, res, body);
      if (data) {
        if (Array.isArray(data)) {
          // data is an array: emit all elements individually
          data.forEach(function (obj) {
            stream.emit('data', obj);
          });
        } else {
          // data is something else: emit as-is
          stream.emit('data', data);
        }
      }

      var nextPage = options.next.call(stream, res, body);
      if (isObject(nextPage)) {
        // nextPage is an object: use as querystring parameters
        Object.keys(nextPage).forEach(function (key) {
          options.qs[key] = nextPage[key];
        });
      } else if (typeof nextPage === 'string') {
        if (isUrl(nextPage)) {
          // a full URL (starting with http:)
          options.uri = parseUrl(nextPage);
        } else {
          // resolve the URL
          options.uri = parseUrl(resolveUrl(options.uri.href, nextPage));
        }
      } else {
        // no nextPage, next iteration will end the stream
        options.uri = null;
      }

      callback();
    });
  });

};

// http compression support for request
// borrowed from https://gist.github.com/nickfishman/5499763
function requestWithEncoding(options, callback) {
  var req = request(options);
  req.on('response', function (res) {
    var chunks = [];
    res.on('data', function (chunk) {
      chunks.push(chunk);
    }).on('end', function () {
      var buffer = Buffer.concat(chunks);
      var encoding = res.headers['content-encoding'];
      if (encoding === 'gzip') {
        zlib.gunzip(buffer, function (err, decoded) {
          callback(err, res, decoded && JSON.parse(decoded.toString()));
        });
      } else if (encoding === 'deflate') {
        zlib.inflate(buffer, function (err, decoded) {
          callback(err, res, decoded && JSON.parse(decoded.toString()));
        });
      } else {
        callback(null, res, JSON.parse(buffer.toString()));
      }
    });
  });
  req.on('error', function (err) {
    callback(err);
  });
}

module.exports = restCollection;
